algosdk/source_map.py

```diff
@@ -0,0 +1,97 @@
+from typing import Dict, Any, List, Tuple
+
+
+class SourceMap:
+    """
+    Decodes a VLQ-encoded source mapping between PC values and TEAL source code lines.
+    Spec available here: https://sourcemaps.info/spec.html
+
+    Args:
+        source_map (dict(str, Any)): source map JSON from algod
+        delimiter (str, optional): delimiter for mappings
+    """
+
+    def __init__(self, source_map: Dict[str, Any], delimiter: str = ";"):
+        self.delimter = delimiter
+
+        self.version: int = source_map["version"]
+        self.sources: List[str] = source_map["sources"]
+        self.mapping: str = source_map["mapping"]
+
+        pc_list = [
+            _decode_int_value(raw_val)
+            for raw_val in self.mapping.split(delimiter)
+        ]
+
+        # Initialize with 0,0 for pc/line
+        self.pc_to_line: Dict[int, int] = {0: 0}
+        self.line_to_pc: Dict[int, List[int]] = {0: [0]}
+
+        last_line = 0
+        for index, line_num in enumerate(pc_list):
+            if line_num is not None:  # be careful for '0' checks!
+                if line_num not in self.line_to_pc:
+                    self.line_to_pc[line_num] = []
+                self.line_to_pc[line_num].append(index)
+                last_line = line_num
+
+            self.pc_to_line[index] = last_line
+
+    def get_line_for_pc(self, pc: int) -> int:
+        return self.pc_to_line[pc]
+
+    def get_pcs_for_line(self, line: int) -> List[int]:
+        return self.line_to_pc[line]
+
+
+def _decode_int_value(value: str) -> int:
+    # Mappings may have up to 5 segments:
+    # Third segment represents the zero-based starting line in the original source represented.
+    decoded_value = base64vlq_decode(value)
+    return decoded_value[2] if decoded_value else None
+
+
+"""
+Source taken from: https://gist.github.com/mjpieters/86b0d152bb51d5f5979346d11005588b
+"""
+
+_b64chars = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
+_b64table = [None] * (max(_b64chars) + 1)
+for i, b in enumerate(_b64chars):
+    _b64table[b] = i
+
+_shiftsize, _flag, _mask = 5, 1 << 5, (1 << 5) - 1
+
+
+def base64vlq_decode(vlqval: str) -> Tuple[int]:
```

Comment:
@barnjamin Minor:  Do you expect the VLQ decode + encode methods to be used _directly_ by SDK users?

My read is:  No - The methods are intended for internal API usage.  If that's the case, propose prefixing these methods with `_` to demarcate as private.
* `base64vlq_decode`
* `base64vlq_encode`

Aside:  If intended for internal usage, it might be possible to remove `base64vlq_encode`.  It's unused.

Reply:
I don't expect they'll be used outside the source map stuff, I'll make them private. 

I am a fan of deleting code but this is somewhat hairy to re-implement without a reference so in this case I'd be in favor of leaving it in unless you've got a strong opinion

Reply:
@barnjamin I agree with your rationale.

I am surprised there's no _well-supported_ reference implementation in Python.  Feels preferable to depend on a library here, but I didn't find one.